torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -68,95 +69,78 @@ def foreach_all_gather(
+        # This cat implicitly casts to `all_gather_input.dtype` (if needed)
+        torch.cat(shards_to_cat, out=all_gather_input)
         all_gather_copy_in_event = torch.cuda.Event()
         all_gather_copy_in_event.record()
     all_gather_stream.wait_event(all_gather_copy_in_event)
     with torch.cuda.stream(all_gather_stream):
         # - All-gather
         all_gather_work = dist.all_gather_into_tensor(
-            output_tensor=flat_tensor,
-            input_tensor=input_flat_tensor_shard,
+            output_tensor=all_gather_output,
+            input_tensor=all_gather_input,
             group=group,
             async_op=async_op,
         )
         all_gather_event = torch.cuda.Event()
         all_gather_event.record()
-        return AllGatherResult(flat_tensor_shards, all_gather_event, all_gather_work)
+        return AllGatherResult(all_gather_output, all_gather_event, all_gather_work)
 
 
 def foreach_all_gather_copy_out(
-    flat_tensor_shards: List[torch.Tensor],
+    all_gather_output: torch.Tensor,  # 1D
     fsdp_params: List[FSDPParam],
     group: dist.ProcessGroup,
     use_uint8: bool,
 ) -> None:
     world_size = group.size()
-    flat_tensor_offset = 0  # [0, total_sharded_numel - 1]
-    for fsdp_param in fsdp_params:
-        param_shards: List[torch.Tensor] = []
-        padded_sharded_numel = fsdp_param.all_gather_input_numel
-        dtype_size = fsdp_param.unsharded_param_data_dtype.itemsize
-        for rank in range(world_size):
-            param_shard_start = rank * padded_sharded_numel
-            numel_in_shard = min(
-                fsdp_param._unsharded_size.numel() - param_shard_start,
-                padded_sharded_numel,
-            )
-            if use_uint8:
-                # Multiply by the dtype size to convert to uint8 numel
-                numel_in_shard *= dtype_size
-            if numel_in_shard > 0:
-                param_shard = flat_tensor_shards[rank].narrow(
-                    0, flat_tensor_offset, numel_in_shard
-                )
-                if use_uint8:
-                    # View back to the expected dtype before the cat
-                    param_shard = param_shard.view(
-                        fsdp_param.unsharded_param_data_dtype
-                    )
-                param_shards.append(param_shard)
-        offset_increment = padded_sharded_numel
-        if use_uint8:
-            offset_increment *= dtype_size
-        flat_tensor_offset += offset_increment
+    if use_uint8:
+        split_sizes_per_rank = [
+            fsdp_param.all_gather_input_numel
+            * fsdp_param.unsharded_param_data_dtype.itemsize
+            for fsdp_param in fsdp_params
+        ]
+    else:
+        split_sizes_per_rank = [
+            fsdp_param.all_gather_input_numel for fsdp_param in fsdp_params
+        ]
+    split_sizes = split_sizes_per_rank * world_size
+    splits = torch.split(all_gather_output, split_sizes, dim=0)
```

Comment:
The largest contributor to CPU overhead reduction was replacing Python `narrow()` calls with a single `torch.split()` to avoid repeated Python-C++ context switches.